torch/lib/c10d/Utils.cpp

```diff
@@ -231,6 +241,16 @@ int connect(
         if (!wait || (!anyRefused && !anyReset)) {
           throw;
         }
+
+        // if a timeout is specified, check time elapsed to see if we need to
+        // timeout. A timeout is specified if timeout != kNoTimeout.
+        if (timeout != kNoTimeout) {
+          const auto elapsed =
+              std::chrono::high_resolution_clock::now() - start;
+          if (elapsed > timeout) {
+            throw std::runtime_error("connect() timed out");
+          }
+        }
         std::this_thread::sleep_for(std::chrono::seconds(1));
         anyRefused = false;
         anyReset = false;
```

Comment:
@pietern Why do we continue retrying here if we have already exhausted all addresses?

Reply:
The process that is supposed to start the TCP initialization daemon may not yet have started and then the host will reply with `ECONNREFUSED`. Somewhat similar, if the daemon has started, but 1000 processes connect at the same time (since they are typically all started at the same time), they may exhaust the listen backlog, and if full, the host replies with a RST packet (hence `ECONNRESET`). In both cases, we should retry until the timeout hits.